### ISSUE TITLE
Python lacks _scproxy on macOS

### ISSUE THREAD

**Issue by healther** [spack/spack#2230]
I'm on a MacOS machine running Sierra and I ran into the problem of a broken spack after loading a spack-installed py2.7.12 .

I gathered that _scproxy is a Mac-only package, although I'm not sure what exactly it is needed for. Is this a known problem? If so where/how can I find the solution to it? 

Any help would be greatly appreciated.

Full traceback:
```
dhcp659-213:git healther$ spack load --dependencies python@2.7.12
dhcp659-213:git healther$ spack
Traceback (most recent call last):
  File "/Users/healther/Documents/git/spack/bin/spack", line 99, in <module>
    import spack
  File "/Users/healther/Documents/git/spack/lib/spack/spack/__init__.py", line 74, in <module>
    import spack.repository
  File "/Users/healther/Documents/git/spack/lib/spack/spack/repository.py", line 45, in <module>
    import spack.spec
  File "/Users/healther/Documents/git/spack/lib/spack/spack/spec.py", line 114, in <module>
    import spack.store
  File "/Users/healther/Documents/git/spack/lib/spack/spack/store.py", line 47, in <module>
    import spack.config
  File "/Users/healther/Documents/git/spack/lib/spack/spack/config.py", line 57, in <module>
    import jsonschema
  File "/Users/healther/Documents/git/spack/lib/spack/external/jsonschema/__init__.py", line 12, in <module>
    from jsonschema.exceptions import (
  File "/Users/healther/Documents/git/spack/lib/spack/external/jsonschema/exceptions.py", line 6, in <module>
    from jsonschema import _utils
  File "/Users/healther/Documents/git/spack/lib/spack/external/jsonschema/_utils.py", line 6, in <module>
    from jsonschema.compat import str_types, MutableMapping, urlsplit
  File "/Users/healther/Documents/git/spack/lib/spack/external/jsonschema/compat.py", line 28, in <module>
    from urllib import unquote  # noqa
  File "/Users/healther/Documents/git/spack/opt/spack/darwin-sierra-x86_64/gcc-4.9.3/python-2.7.12-iptjvhtegkxoyzhborfkjkndx4xnaerr/lib/python2.7/urllib.py", line 1431, in <module>
    from _scproxy import _get_proxy_settings, _get_proxies
ImportError: No module named _scproxy
```

**Comment by citibeth** [comment]
This looks like a problem with Spack's macOS installation of Python.  See:

http://stackoverflow.com/questions/5647023/no-module-named-scproxy-on-osx

**Comment by citibeth** [comment]
The workaround for now would be to hard-code the Python used for your Spack.  Edit the file `spack/bin/spack`.  Change the first line from:
```
#!/usr/bin/env python
```
to:
```
#!/usr/bin/python
```
(or wherever the System python lives on your machine).

**Comment by healther** [comment]
Thanks for the quick reply, that fixed my immediate problem. However it won't help for libraries that explicitly need _scproxy, as is apparently the case for NEST (my initial installation target)

Any suggestions on how I can provide the module?

**Comment by citibeth** [comment]
Building stuff for Mac is problematic, in my experience.  Things don't "just work."  Starting with Yosemite, many things won't build with GCC; and yet Apple offers no Fortran compiler.  And little minefields abound.

How to build the module?  Find instructions on-line somewhere and follow them, thus creating a Spack package.  Test it make sure it works, then add it as a macOS-only dependency to the appropriate Spack packages.  Then submit a PR back to Spack.

I dumped macOS early this year so I wouldn't have to do that.  Others may be able to give more help @davydden

**Comment by tgamblin** [comment]
we should probably just make the mac build [do this](http://stackoverflow.com/questions/5647023/no-module-named-scproxy-on-osx)

**Comment by davydden** [comment]
i don't use python much, i never came across this problem, sorry.

**Comment by healther** [comment]
Thanks for your comments, I'll take a look at the effort that _scproxy would require, but I feel I'll have to get some kind of native Linux installation for a somewhat pain free solution

**Comment by citibeth** [comment]
Yes, native Linux was my conclusion.  I reformatted my HDD and had CentOS 7
installed (the Linux allowed at our site).  I also know people who run a
Linux VM on top of macOS.

**Comment by tgamblin** [comment]
I'm actually unable to reproduce this on elcapitan.  @healther: what compiler are you using?  Is your `gcc@4.9.3` from Spack or something else?

**Comment by healther** [comment]
Sorry for the slow reply, I was away over the weekend. I'm on Sierra (10.12.1), not sure it that makes the difference.

Yes I'm using a gcc@4.9.3 from spack built by an gcc@6.2.0 also from spack but built by the native clang (I guess its comes with Xcode).
Can you actually use urllib (according to the SO article from above it actually uses _scproxy)?

Not sure if this is helpful information:
Right now I'm trying to implement the workaround mentioned on SO, but I'm a bit confused there, as to why that would make any difference. If I understand it correctly spack should do exactly what the proposed solution does (i.e. building a clean python from python.org). And then injecting the _scproxy.so into the native python.
When I'm manually building the code from the Python-2.7.12.tgz I get a ```_scproxy.so``` file under ```./build/lib.macosx-10.4-x86-2.7/```, however I did not yet manage to use the gcc@4.9.3 from spack for that build...

**Comment by citibeth** [comment]
Python does not build with gcc on recent macs, in my experience.  Must use
clang.

**Comment by healther** [comment]
Spack managed to build python with gcc. At least I see
```
healther$ spack find python
==> 2 installed packages.
-- darwin-sierra-x86_64 / gcc@4.9.3 -----------------------------
python@2.7.11  python@2.7.12
healther$ spack load --dependencies python@2.7.12
healther$ python
Python 2.7.12 (default, Nov  3 2016, 14:15:50) 
[GCC 4.9.3] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```

So python claims to have been build by a gcc@4.9.3. On the other hand, I fail to do this manually... I.e. trying to configure the python source code with ```CC=gcc && CXX=g++``` fails due to an t_size error.

However following the instructions from [SO](http://stackoverflow.com/questions/5647023/no-module-named-scproxy-on-osx) (except using the 2.7.12 instead of an 2.6 python) and copy/pasting the ```_scproxy.so``` from the clang build folder into the spack python@2.7.12-folder ```$SPACKHOME/opt/spack/darwin-sierra-x86_64/gcc-4.9.3/python-2.7.12-iptjvhtegkxoyzhborfkjkndx4xnaerr/lib/python2.7/lib-dynload/``` prevents the call ```import urllib``` from failing. 

I'll keep an eye out whether this will produce a problem at some point... (If I understand what happened correctly, I now use a _scproxy.so which is build by a different compiler than the one that build the python binary, right?)

**Comment by citibeth** [comment]
On Mon, Nov 7, 2016 at 7:47 AM, healther <notifications@github.com> wrote:

> Spack managed to build python with gcc. At least I see
>
> healther$ spack find python
> ==> 2 installed packages.
> -- darwin-sierra-x86_64 / gcc@4.9.3 -----------------------------
> python@2.7.11  python@2.7.12
> healther$ spack load --dependencies python@2.7.12
> healther$ python
> Python 2.7.12 (default, Nov  3 2016, 14:15:50)
> [GCC 4.9.3] on darwin
> Type "help", "copyright", "credits" or "license" for more information.
> >>>
>
> So python claims to have been build by a gcc@4.9.3. On the other hand, I
> fail to do this manually... I.e. trying to configure the python source code
> with CC=gcc && CXX=g++ fails due to an t_size error.
>

I would look at `python/package.py`, this should tell you what Spack is
doing to build Python.


> However following the instructions from SO
> <http://stackoverflow.com/questions/5647023/no-module-named-scproxy-on-osx>
> (except using the 2.7.12 instead of an 2.6 python) and copy/pasting the
> _scproxy.so from the clang build folder into the spack
> python@2.7.12-folder $SPACKHOME/opt/spack/darwin-sierra-x86_64/gcc-4.9.3/
> python-2.7.12-iptjvhtegkxoyzhborfkjkndx4xnaerr/lib/python2.7/lib-dynload/
> prevents the call import urllib from failing.
>
That's a good start... any way this can be incorporated into
`python/package.py`?


> I'll keep an eye out whether this will produce a problem at some point...
> (If I understand what happened correctly, I now use a _scproxy.so which is
> build by a different compiler than the one that build the python binary,
> right?)
>
It's OK to mix-n-match compilers with C code, but not C++ or Fortran.

**Comment by healther** [comment]
Okay I found a setting that works, but I don't know how I can put that into the ```python/package.py```. 
Doing
```
make clean
spack load gcc@4.9.3
MACOSX_DEPLOYMENT_TARGET=10.6 ./configure --prefix=$PWD/build00 CC=gcc CXX=g++ --disable-toolbox-glue
make
make install
spack unload gcc
./configure --prefix=$PWD/build01
make sharedmods
find . -name '_scproxy.so' -exec cp {} build00/lib/python2.7/lib-dynload/ \;
```
in the 2.7.12 source code folder, gives a working (i.e. ```import urllib``` does not fail) python binary in ```build00/bin/python```.

Essentially one needs to additionally build the sharedmods with the system compiler and without the OSX specific flags (i.e. toolboxglue and DEPLOYMENT_TARGET).

In order to incorporate this into the ```python/package.py``` one would need to add the second calls to ```./configure``` and ```make``` after the ones currently called and then moving the ```_scproxy.so``` into the . However I don't know how I would do the ```unload gcc``` from within spack and how I would find the build/install paths of the spack installed python.
What would be the canonical way to decide whether to use this hacky solution? Since it is working on ElCapitan @tgamblin?

**Comment by citibeth** [comment]
One possible way would be the rare case of two Spack packages downloading
the same tarball.

`scproxy/package.py` would be built with clang, and install just
`_scproxy.so` in its install tree. `python/package.py` would be built with
GCC, and depend on `scproxy%clang` as a build dependency.  It would copy
the `_scproxy.so` that was already built into its install tree.

**Comment by healther** [comment]
Has there been any progress on this? I'm unfortunately still forced to use a gcc-build python

**Comment by LDAmorim** [comment]
Hi,
Sorry to bother you again with such an old issue, but I am new to spack and can't solve this issue.
I compiled gcc@9.1.0 and python@3.7.0 (with %gcc) with spack successfully on MacOS Mojave.
Now I was trying to install python packages (f.eg. matplotlib) with spack.
When I try to install them, I get the error message `ModuleNotFoundError: No module named '_scproxy'` on the final line in the spack-build-out.txt file.
Do you know if it is still not possible to install python and its packages in Mac with spack?
Thanks,

**Comment by citibeth** [comment]
Things work on Mac, as long as you use the Mac-supplied clang compiler.  There is stuff in the Spack docs on mixing Clang + gfortran to get a full compiler stack.

**Comment by LDAmorim** [comment]
Hi, Thanks! I will try to use then both clang + gfortran.

**Comment by citibeth** [comment]
I'm closing this because I don't think it will ever be realistic to build significant software stacks on macOS with GCC.

**Comment by hartzell** [comment]
> There is stuff in the Spack docs on mixing Clang + gfortran to get a full compiler stack.

I've found this [Ubuntu specific bit of doc for clang+gofortran](https://spack-tutorial.readthedocs.io/en/latest/tutorial_configuration.html#compiler-configuration).

Is there anything Darwin specific hidden somewhere?  Specifically, does it matter where I get the gfortran that I mix in with the Apple supplied clang?

I'm working on a mac running 10.13.6, using gcc@9.2.0 from homebrew.

I just hit this with a new package that I'm putting together that dies with:

```
ModuleNotFoundError: No module named '_scproxy'
```

while trying to build py-setuptools.

**Comment by LDAmorim** [comment]
Just in case anyone is still looking for possible workarounds to this issue, here is an alternative in the same liens.
In current macOS systems there is an inbuilt python so it is possible to just copy its `_scproxy` file. So there is no need to do an additional python installation, even if using `gcc` on macOS.
In my case the following worked:
```console
$ cp /Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/lib-dynload/_scproxy.cpython-37m-darwin.so /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-skylake/gcc-9.3.0/python-3.7.6-rztnajc6br75odyhxikw7hfk52fnddb7/lib/python3.7/lib-dynload/
$ spack install py-setuptools %gcc
```

**Comment by ax3l** [comment]
Since this problem persists and is not python2 specific, let's reopen this one.

Related GCC 8.3.0 build error on macOS: https://bugs.python.org/issue26317#msg342055

Maybe passing `--enable-framework=${HOME}/sandbox/src/python3.7/inst` could work for the CPython install?

```
$ ./configure --help
# ...
  --disable-framework
  --enable-framework[=INSTALLDIR]
                          Build (MacOSX|Darwin) framework
# ...
  --with-framework-name=FRAMEWORK
                          specify an alternate name of the framework built
                          with --enable-framework
```

we already set `PYTHONFRAMEWORKPREFIX` but this might not be enough to successfully build and install `_scproxy` with GCC. Or we just set `--disable-framework`, if that works.

Look what Python ~~3.7.7~~ 3.7.0 fixed:
- https://docs.python.org/3.7/whatsnew/changelog.html#python-3-7-7-final
  - `bpo-28440`: No longer add /Library/Python/3.x/site-packages to sys.path for macOS framework builds to avoid future conflicts. (3.7.0+)
  - https://bugs.python.org/issue28440

And the macOS-fixed Python 3.7.7 landed in Spack yesterday #16170 (thx @adamjstewart @michaelkuhn), so maybe all above can give it a try?
```
commit 45fee6edbd696ae1180dda066a8e714bc8676349
Author: Adam J. Stewart @adamjstewart 
Date:   Mon Apr 20 16:57:08 2020 -0500

    Python: add latest Python 2 releases (#16191)

commit 1f49fb8cebe5dfe1ca2f5caeb614f266dc5912e6
Author: Michael Kuhn @michaelkuhn 
Date:   Mon Apr 20 22:32:58 2020 +0200

    python: Add 3.7.7 (#16170)
```

Update: hm, @LDAmorim already builds with 3.7.6... (and the patch is 3.7.0+). So maybe `--disable-framework` after all?

Still failing on yhe first package that uses CPython with `python@3.7.7`: https://github.com/spack/macos-nightly/runs/606624706?check_suite_focus=true

**Comment by adamjstewart** [comment]
Does anyone have any other suggestions to fix this? I know I maintain Spack's Python recipe, but I've never been able to reproduce this bug, so it's hard to fix.

**Comment by ax3l** [comment]
This is actually an upstream problem that occurs when installing (C)Python with GCC on macOS.
There is a silent build bug that does not build the module in questions because it includes objective-C code from system libs.

This has recently been addressed but upstream added no CI for it. Here are the upstream reports for this:
- https://bugs.python.org/issue26317
- https://github.com/python/cpython/pull/13306 (proposed)
- https://github.com/python/cpython/pull/20176 (merged but reverted due to ongoing issues on FreeBSD)
- https://github.com/python/cpython/pull/19681 (unmerged CI coverage)

### TL;DR
- only macOS + GCC issue
  - was scheduled for the upcoming CPython 3.9+ but [status unclear to me](https://github.com/python/cpython/pull/20176#issuecomment-679432592)
- or just build with AppleClang on macOS

**Comment by ax3l** [comment]
I added a conflict a while back for this in #16249

This is an upstream issue and upstream has currently little capacity to ensure any compiler but the system compiler (AppleClang) works.
